PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp

```diff
@@ -1208,10 +1787,246 @@ int FileRelocalizerDlg::doDialog(const TCHAR *fn, bool isRTL)
+{
+	TCHAR textBuffer[MAX_PATH];
+	TVITEM tvItem;
+	tvItem.mask = TVIF_TEXT | TVIF_PARAM;
+	tvItem.pszText = textBuffer;
+	tvItem.cchTextMax = MAX_PATH;
+
+	tvItem.hItem = hItem;
+	SendMessage(_treeView->getHSelf(), TVM_GETITEM, 0,reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
+	assert (data.isBaseDirectory() || data.isDirectory());
+	if (!data.isBaseDirectory() && !data.isDirectory())
+		return;
+	_exists = PathFileExists(data._filePath.c_str()) != 0;
+	if (_exists)
+		_path = data._filePath;
+
+    for (HTREEITEM hItemNode = _treeView->getChildFrom(hItem);
+		hItemNode != NULL;
+		hItemNode = _treeView->getNextSibling(hItemNode))
+	{
+		tvItem.hItem = hItemNode;
+		SendMessage(_treeView->getHSelf(), TVM_GETITEM, 0,reinterpret_cast<LPARAM>(&tvItem));
+		ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
+
+		generic_string filename(data._filePath);
+		const size_t lastSlashIdx = filename.find_last_of(TEXT("\\/"));
+		if (std::string::npos != lastSlashIdx)
+			filename.erase(0, lastSlashIdx + 1);
+
+		if (data.isDirectoryFile())
+		{
+			_files.insert(filename);
+			_fileMap[filename] = hItemNode;
+		}
+		else if (data.isBaseDirectory() || data.isDirectory())
+		{
+			_dirs.insert(filename);
+			_dirMap[filename] = hItemNode;
+		}
+		else
+		{
+			assert(0);
+		}
+	}
+
+}
+
+
+void ProjectPanel::TreeUpdaterDirectory::onDirAdded(const generic_string& name)
+{
+	_projectPanel->addDirectory(_hItem, name.c_str(), false, (_path + TEXT("\\") + name).c_str() );
+
+}
+
+void ProjectPanel::TreeUpdaterDirectory::onDirRemoved(const generic_string& name)
+{
+	_treeView->removeItem(_dirMap[name]);
+}
+
+void ProjectPanel::TreeUpdaterDirectory::onFileAdded(const generic_string& name)
+{
+	_treeView->addItem(name.c_str(), _hItem, INDEX_LEAF_DIRFILE, new ProjectPanelData(_projectPanel->_directoryWatcher, name.c_str(), (_path + TEXT("\\") + name).c_str(), nodeType_dirFile ));
+
+}
+
+void ProjectPanel::TreeUpdaterDirectory::onFileRemoved(const generic_string& name)
+{
+	_treeView->removeItem(_fileMap[name]);
+}
+
+int CALLBACK compareFunc(LPARAM lhs, LPARAM rhs, LPARAM)
+{
+	ProjectPanel::ProjectPanelData& dataL = *reinterpret_cast<ProjectPanel::ProjectPanelData*>(lhs);
+	ProjectPanel::ProjectPanelData& dataR = *reinterpret_cast<ProjectPanel::ProjectPanelData*>(rhs);
+	assert( (dataL.isDirectoryFile() || dataL.isDirectory()) && (dataR.isDirectoryFile() || dataR.isDirectory()));
+	if( !((dataL.isDirectoryFile() || dataL.isDirectory()) && (dataR.isDirectoryFile() || dataR.isDirectory())))
+		return 0;
+
+	// one of the items is a folder, the other is not - folders are always on top
+	if (dataL.isDirectory() != dataR.isDirectory())
+		return dataL.isDirectory() ? -1 : 1;
+
+	// both items are of the same kind.
+	return lstrcmpi(dataL._name.c_str(),dataR._name.c_str());
+}
+
+void ProjectPanel::TreeUpdaterDirectory::onEndSynchronize(const Directory& other)
+{
+	other;

```

Comment:
This is just a way to overcome
"c4100 unreferenced formal parameter".
I prefer this way, since it clearly shows that this parameter is not used, without removing its meaning by removing the name from the parameter list.

Anyway, I do not see the usefulness of this warning at all. It makes things far less readable without any real benefit (especially in optional virtual functions). I suggest to switch it off globally.

But of course this is not my project ;), I can change it to removing the name(s) from the parameter list(s), if you like this better.